File: js/src/Ice/HashUtil.js
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

const Ice = require("../Ice/StringUtil").Ice;

const StringUtil = Ice.StringUtil;

class HashUtil
{
    static addBoolean(h, b)
    {
        return ((h << 5) + h) ^ (b ? 0 : 1);
    }

    static addString(h, str)
    {
        if(str !== undefined && str !== null)
        {
            h = ((h << 5) + h) ^ StringUtil.hashCode(str);
        }
        return h;
    }

    static addNumber(h, num)
    {
        return ((h << 5) + h) ^ num;
    }

    static addHashable(h, obj)
    {
        if(obj !== undefined && obj !== null)
        {
            h = ((h << 5) + h) ^ obj.hashCode();
        }
        return h;
    }

    static addArray(h, arr, hashCode)
    {
        if(arr !== undefined && arr !== null)
        {
            for(let i = 0; i < arr.length; ++i)
            {
                h = hashCode(h, arr[i]);
            }
        }
        return h;
    }
}

Ice.HashUtil = HashUtil;
module.exports.Ice = Ice;

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/HashUtil.java
// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package com.zeroc.IceInternal;

public final class HashUtil
{
    public static int
    hashAdd(int hashCode, boolean value)
    {
        return ((hashCode << 5) + hashCode) ^ (value ? 0 : 1);
    }

    public static int
    hashAdd(int hashCode, short value)
    {
        return ((hashCode << 5) + hashCode) ^ (int)(2654435761l * value);
    }

    public static int
    hashAdd(int hashCode, byte value)
    {
        return ((hashCode << 5) + hashCode) ^ (int)(2654435761l * value);
    }

    public static int
    hashAdd(int hashCode, int value)
    {
        return ((hashCode << 5) + hashCode) ^ (int)(2654435761l * value);
    }

    public static int
    hashAdd(int hashCode, long value)
    {
        return ((hashCode << 5) + hashCode) ^ (int)(value ^ (value >>> 32));
    }

    public static int
    hashAdd(int hashCode, float value)
    {
        return ((hashCode << 5) + hashCode) ^ Float.floatToIntBits(value);
    }

    public static int
    hashAdd(int hashCode, double value)
    {
        long v = Double.doubleToLongBits(value);
        return ((hashCode << 5) + hashCode) ^ (int)(v ^ (v >>> 32));
    }

    public static int
    hashAdd(int hashCode, java.lang.Object value)
    {
        if(value != null)
        {
            hashCode = ((hashCode << 5) + hashCode) ^ value.hashCode();
        }
        return hashCode;
    }

    public static int
    hashAdd(int hashCode, boolean[] arr)
    {
        return ((hashCode << 5) + hashCode) ^ java.util.Arrays.hashCode(arr);
    }

    public static int
    hashAdd(int hashCode, byte[] arr)
    {
        return ((hashCode << 5) + hashCode) ^ java.util.Arrays.hashCode(arr);
    }

    public static int
    hashAdd(int hashCode, char[] arr)
    {
        return ((hashCode << 5) + hashCode) ^ java.util.Arrays.hashCode(arr);
    }

    public static int
    hashAdd(int hashCode, double[] arr)
    {
        return ((hashCode << 5) + hashCode) ^ java.util.Arrays.hashCode(arr);
    }

    public static int
    hashAdd(int hashCode, float[] arr)
    {
        return ((hashCode << 5) + hashCode) ^ java.util.Arrays.hashCode(arr);
    }

    public static int
    hashAdd(int hashCode, int[] arr)
    {
        return ((hashCode << 5) + hashCode) ^ java.util.Arrays.hashCode(arr);
    }

    public static int
    hashAdd(int hashCode, long[] arr)
    {
        return ((hashCode << 5) + hashCode) ^ java.util.Arrays.hashCode(arr);
    }

    public static int
    hashAdd(int hashCode, Object[] arr)
    {
        return ((hashCode << 5) + hashCode) ^ java.util.Arrays.hashCode(arr);
    }

    public static int
    hashAdd(int hashCode, short[] arr)
    {
        return ((hashCode << 5) + hashCode) ^ java.util.Arrays.hashCode(arr);
    }
}
